include/picongpu/particles/collision/detail/ListEntry.hpp

```diff
@@ -0,0 +1,292 @@
+                                    break;
+                                }
+                                else
+                                {
+#ifndef BOOST_COMP_HIP
+                                    // TODO: took it from ParticleBox.hpp but HIP should support printf, though
+                                    // maby only with HCC_ENABLE_PRINTF set.
+                                    printf(
+                                        "%s in collisions:  mallocMC out of memory (try %i of %i)\n",
+                                        (numTries + 1) == maxTries ? "ERROR" : "WARNING",
+                                        numTries + 1,
+                                        maxTries);
+#endif
+                                }
+                            }
+                            if(ptrToIndicies == nullptr)
+                            {
+#if BOOST_COMP_HIP
+                                // do nothing
+#elif BOOST_LANG_CUDA
+                                __trap();
+#else
+                                throw std::runtime_error("Out of memory in collisions.");
+#endif
+                            }
+                        }
+                        // reset counter
+                        size = 0u;
+                    }
+
+                    /* Release allocated heap memory.
+                     *
+                     * @param acc alpaka accelerator
+                     * @param deviceHeapHandle Heap handle used for allocating storage on device.
+                     */
+                    template<typename T_acc, typename T_DeviceHeapHandle>
+                    DINLINE void finalize(T_acc const& acc, T_DeviceHeapHandle& deviceHeapHandle)
+                    {
+                        if(ptrToIndicies != nullptr)
+                        {
+#if(PMACC_CUDA_ENABLED == 1)
+                            deviceHeapHandle.free(acc, (void*) ptrToIndicies);
+                            ptrToIndicies = nullptr;
+#else
+                            delete(ptrToIndicies);
+#endif
+                        }
+                    }
+
+
+                    /* Shuffle list entries.
+                     *
+                     * @param acc alpaka accelerator
+                     * @param rngHandle random number generator handle
+                     */
+                    template<typename T_Acc, typename T_RngHandle>
+                    DINLINE void shuffle(T_Acc const& acc, T_RngHandle& rngHandle)
+                    {
+                        using UniformUint32_t = pmacc::random::distributions::Uniform<uint32_t>;
+                        auto rng = rngHandle.template applyDistribution<UniformUint32_t>();
+                        // shuffle the particle lookup table
+                        for(uint32_t i = size; i > 1; --i)
+                        {
+                            /* modulo is not perfect but okish,
+                             * because of the loop head mod zero is not possible
```

Comment:
It's  actually @psychocoderHPC's comment. Maybe he can say sth more.

Reply:
I think I also asked about it a while ago offline, thinking it's a bad way to shuffle. But I think i was just wrong and this is fine.

Reply:
The first line helped me to at least guess the random method, but I was totally confused by the second line. @psychocoderHPC what did you want to say here?

Reply:
@PrometheusPi @sbastrakov I guess it only means that we don't divide by zero in `p=rng(acc) % i;` because of the `i>1` condition.

Reply:
`i > 1` is required because we access `ptrToIndicies[i - 1]`, without the guard it will be an out of memory access.
`rng(acc) % i` modulo should be not used if you like to keep the normal distribution of your random numbers. This should not be critical at this point.